File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/DiscoverPeopleAdapter.java
package com.veganbeauty.app.features.community.com_feed;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


import coil.Coil;
import coil.ImageLoader;
import coil.decode.SvgDecoder;
import coil.request.ImageRequest;

import com.veganbeauty.app.R;
import com.veganbeauty.app.data.local.entities.UserEntity;
import com.veganbeauty.app.databinding.ComItemDiscoverPersonBinding;

import java.util.List;

public class DiscoverPeopleAdapter extends RecyclerView.Adapter<DiscoverPeopleAdapter.PersonViewHolder> {

    public enum ActionType {
        SUGGEST, REQUEST, FOLLOW_BACK
    }

    public interface OnActionClickListener {
        void onActionClick(UserEntity user, String action);
    }

    private List<UserEntity> users;
    private final ActionType actionType;
    private final OnActionClickListener onActionClick;

    public DiscoverPeopleAdapter(List<UserEntity> users, ActionType actionType, OnActionClickListener onActionClick) {
        this.users = users;
        this.actionType = actionType;
        this.onActionClick = onActionClick;
    }

    public static class PersonViewHolder extends RecyclerView.ViewHolder {
        public final ComItemDiscoverPersonBinding binding;

        public PersonViewHolder(ComItemDiscoverPersonBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    @NonNull
    @Override
    public PersonViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ComItemDiscoverPersonBinding binding = ComItemDiscoverPersonBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new PersonViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull PersonViewHolder holder, int position) {
        UserEntity user = users.get(position);
        holder.binding.tvUsername.setText(user.getUsername());

        if (user.getAvatar() != null && !user.getAvatar().isEmpty()) {
            ImageLoader imageLoader = Coil.imageLoader(holder.binding.ivAvatar.getContext());
            ImageRequest request = new ImageRequest.Builder(holder.binding.ivAvatar.getContext())
                    .data(user.getAvatar())
                    .decoderFactory(new SvgDecoder.Factory())
                    .placeholder(android.R.color.darker_gray)
                    .error(R.drawable.img_avatar)
                    .target(holder.binding.ivAvatar)
                    .build();
            imageLoader.enqueue(request);
        } else {
            holder.binding.ivAvatar.setImageResource(android.R.color.darker_gray);
        }

        if (user.getMutualCount() > 0) {
            String friendName = user.getFirstMutualFriendName() != null ? user.getFirstMutualFriendName() : "Ai đó";
            if (user.getMutualCount() == 1) {
                holder.binding.tvSubtitle.setText("Có " + friendName + " đang theo dõi");
            } else {
                holder.binding.tvSubtitle.setText("Có " + friendName + " và " + (user.getMutualCount() - 1) + " người khác đang theo dõi");
            }
            holder.binding.tvSubtitle.setVisibility(View.VISIBLE);

            List<String> avatars = user.getMutualFriendAvatars();
            if (avatars != null && !avatars.isEmpty()) {
                holder.binding.flMutualAvatars.setVisibility(View.VISIBLE);

                if (avatars.size() >= 1) {
                    holder.binding.cvMutual1.setVisibility(View.VISIBLE);
                    loadImage(holder.itemView.getContext(), avatars.get(0), holder.binding.ivMutual1);
                } else holder.binding.cvMutual1.setVisibility(View.GONE);

                if (avatars.size() >= 2) {
                    holder.binding.cvMutual2.setVisibility(View.VISIBLE);
                    loadImage(holder.itemView.getContext(), avatars.get(1), holder.binding.ivMutual2);
                } else holder.binding.cvMutual2.setVisibility(View.GONE);

                if (avatars.size() >= 3) {
                    holder.binding.cvMutual3.setVisibility(View.VISIBLE);
                    loadImage(holder.itemView.getContext(), avatars.get(2), holder.binding.ivMutual3);
                } else holder.binding.cvMutual3.setVisibility(View.GONE);

            } else {
                holder.binding.flMutualAvatars.setVisibility(View.GONE);
            }
        } else {
            holder.binding.tvSubtitle.setVisibility(View.GONE);
            holder.binding.flMutualAvatars.setVisibility(View.GONE);
        }

        switch (actionType) {
            case SUGGEST:
                holder.binding.btnPrimary.setText("Theo dõi");
                holder.binding.btnSecondary.setVisibility(View.GONE);
                holder.binding.ivClose.setVisibility(View.VISIBLE);
                break;
            case REQUEST:
                holder.binding.btnPrimary.setText("Xác nhận");
                holder.binding.btnSecondary.setVisibility(View.VISIBLE);
                holder.binding.ivClose.setVisibility(View.GONE);
                break;
            case FOLLOW_BACK:
                holder.binding.btnPrimary.setText("Theo dõi lại");
                holder.binding.btnSecondary.setVisibility(View.GONE);
                holder.binding.ivClose.setVisibility(View.VISIBLE);
                break;
        }

        holder.binding.btnPrimary.setOnClickListener(v -> {
            String btnText = holder.binding.btnPrimary.getText().toString();
            if (btnText.equals("Theo dõi") || btnText.equals("Theo dõi lại")) {
                if (onActionClick != null) onActionClick.onActionClick(user, "FOLLOW");
                holder.binding.btnPrimary.setText("Đang theo dõi");
                holder.binding.btnPrimary.setBackgroundResource(R.drawable.com_bg_filter_normal);
                holder.binding.btnPrimary.setTextColor(holder.itemView.getContext().getResources().getColor(R.color.primary, null));
            } else if (btnText.equals("Xác nhận")) {
                if (onActionClick != null) onActionClick.onActionClick(user, "ACCEPT_REQUEST");
                holder.binding.btnPrimary.setText("Đã xác nhận");
                holder.binding.btnSecondary.setVisibility(View.GONE);
            }
        });
    }

    private void loadImage(android.content.Context context, String url, android.widget.ImageView target) {
        ImageLoader imageLoader = Coil.imageLoader(target.getContext());
        ImageRequest request = new ImageRequest.Builder(target.getContext())
                .data(url)
                .decoderFactory(new SvgDecoder.Factory())
                .error(R.drawable.img_avatar)
                .transformations(new coil.transform.CircleCropTransformation())
                .target(target)
                .build();
        imageLoader.enqueue(request);
    }

    @Override
    public int getItemCount() {
        return users.size();
    }

    public void updateData(List<UserEntity> newUsers) {
        users = newUsers;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/DiscoverPeopleAdapter.kt
package com.veganbeauty.app.features.community.com_feed

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import coil.decode.SvgDecoder
import coil.load
import com.veganbeauty.app.R
import com.veganbeauty.app.data.local.entities.UserEntity
import com.veganbeauty.app.databinding.ComItemDiscoverPersonBinding

class DiscoverPeopleAdapter(
    private var users: List<UserEntity>,
    private val actionType: ActionType,
    private val onActionClick: ((UserEntity, String) -> Unit)? = null
) : RecyclerView.Adapter<DiscoverPeopleAdapter.PersonViewHolder>() {

    enum class ActionType {
        SUGGEST, REQUEST, FOLLOW_BACK
    }

    class PersonViewHolder(val binding: ComItemDiscoverPersonBinding) : RecyclerView.ViewHolder(binding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PersonViewHolder {
        val binding = ComItemDiscoverPersonBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return PersonViewHolder(binding)
    }

    override fun onBindViewHolder(holder: PersonViewHolder, position: Int) {
        val user = users[position]
        holder.binding.tvUsername.text = user.username
        
        if (!user.avatar.isNullOrEmpty()) {
            holder.binding.ivAvatar.load(user.avatar) {
                decoderFactory(SvgDecoder.Factory())
                crossfade(true)
                placeholder(android.R.color.darker_gray)
                error(R.drawable.img_avatar)
            }
        } else {
            holder.binding.ivAvatar.setImageResource(android.R.color.darker_gray)
        }
        
        if (user.mutualCount > 0) {
            val friendName = user.firstMutualFriendName ?: "Ai đó"
            if (user.mutualCount == 1) {
                holder.binding.tvSubtitle.text = "Có $friendName đang theo dõi"
            } else {
                holder.binding.tvSubtitle.text = "Có $friendName và ${user.mutualCount - 1} người khác đang theo dõi"
            }
            holder.binding.tvSubtitle.visibility = View.VISIBLE
            
            // Show avatars if available
            val avatars = user.mutualFriendAvatars
            if (avatars.isNotEmpty()) {
                holder.binding.flMutualAvatars.visibility = View.VISIBLE
                
                // Show up to 3 avatars
                if (avatars.size >= 1) {
                    holder.binding.cvMutual1.visibility = View.VISIBLE
                    holder.binding.ivMutual1.load(avatars[0]) {
                        decoderFactory(SvgDecoder.Factory())
                        transformations(coil.transform.CircleCropTransformation())
                        crossfade(true)
                        error(R.drawable.img_avatar)
                    }
                } else holder.binding.cvMutual1.visibility = View.GONE
                
                if (avatars.size >= 2) {
                    holder.binding.cvMutual2.visibility = View.VISIBLE
                    holder.binding.ivMutual2.load(avatars[1]) {
                        decoderFactory(SvgDecoder.Factory())
                        transformations(coil.transform.CircleCropTransformation())
                        crossfade(true)
                        error(R.drawable.img_avatar)
                    }
                } else holder.binding.cvMutual2.visibility = View.GONE
                
                if (avatars.size >= 3) {
                    holder.binding.cvMutual3.visibility = View.VISIBLE
                    holder.binding.ivMutual3.load(avatars[2]) {
                        decoderFactory(SvgDecoder.Factory())
                        transformations(coil.transform.CircleCropTransformation())
                        crossfade(true)
                        error(R.drawable.img_avatar)
                    }
                } else holder.binding.cvMutual3.visibility = View.GONE
                
            } else {
                holder.binding.flMutualAvatars.visibility = View.GONE
            }
        } else {
            holder.binding.tvSubtitle.visibility = View.GONE
            holder.binding.flMutualAvatars.visibility = View.GONE
        }

        // Configure buttons based on ActionType
        when (actionType) {
            ActionType.SUGGEST -> {
                holder.binding.btnPrimary.text = "Theo dõi"
                holder.binding.btnSecondary.visibility = View.GONE
                holder.binding.ivClose.visibility = View.VISIBLE
            }
            ActionType.REQUEST -> {
                holder.binding.btnPrimary.text = "Xác nhận"
                holder.binding.btnSecondary.visibility = View.VISIBLE
                holder.binding.ivClose.visibility = View.GONE
            }
            ActionType.FOLLOW_BACK -> {
                holder.binding.btnPrimary.text = "Theo dõi lại"
                holder.binding.btnSecondary.visibility = View.GONE
                holder.binding.ivClose.visibility = View.VISIBLE
            }
        }

        holder.binding.btnPrimary.setOnClickListener {
            // Mock Action
            if (holder.binding.btnPrimary.text.toString() == "Theo dõi" || holder.binding.btnPrimary.text.toString() == "Theo dõi lại") {
                onActionClick?.invoke(user, "FOLLOW")
                holder.binding.btnPrimary.text = "Đang theo dõi"
                holder.binding.btnPrimary.setBackgroundResource(R.drawable.com_bg_filter_normal)
                holder.binding.btnPrimary.setTextColor(holder.itemView.context.getColor(R.color.primary))
            } else if (holder.binding.btnPrimary.text.toString() == "Xác nhận") {
                onActionClick?.invoke(user, "ACCEPT_REQUEST")
                holder.binding.btnPrimary.text = "Đã xác nhận"
                holder.binding.btnSecondary.visibility = View.GONE
            }
        }
    }

    override fun getItemCount() = users.size

    fun updateData(newUsers: List<UserEntity>) {
        users = newUsers
        notifyDataSetChanged()
    }
}
